bokehjs/src/coffee/models/tools/edit/poly_edit_tool.ts

```diff
@@ -0,0 +1,244 @@
+      let xs = ds.data[xkey][index];
+      if ((xs.concat == null)) {
+        // Convert typed arrays to regular arrays for editing
+        ds.data[xkey][index] = (xs = copy(xs));
+      }
+      if (pxkey) { point_ds.data[pxkey] = xs; }
+    } else {
+      point_glyph.x = {value: glyph.xs.value};
+    }
+    if (ykey) {
+      let ys = ds.data[ykey][index];
+      // Convert typed arrays to regular arrays for editing
+      if ((ys.concat == null)) {
+        ds.data[ykey][index] = (ys = copy(ys));
+      }
+      if (pykey) { point_ds.data[pykey] = ys; }
+    } else {
+      point_glyph.y = {value: glyph.ys.value};
+    }
+    point_ds.selected['1d'].indices = [];
+    this._selected_renderer = renderer;
+    point_ds.change.emit(undefined);
+    point_ds.properties.data.change.emit(undefined);
+  }
+
+  _move(e: BkEv): void {
+    if (this._drawing && this._selected_renderer != null) {
+      const renderer = this.model.vertex_renderer;
+      const point = this._map_drag(e.bokeh.sx, e.bokeh.sy, renderer);
+      if (point == null) { return; }
+      const [x, y] = point;
+      const ds = renderer.data_source;
+      const glyph: any = renderer.glyph;
+      const [xkey, ykey] = [glyph.x.field, glyph.y.field];
+      const index = ds.selected['1d'].indices[0];
+      if (xkey) { ds.data[xkey][index] = x; }
+      if (ykey) { ds.data[ykey][index] = y; }
+      ds.change.emit(undefined);
+      this._selected_renderer.data_source.change.emit(undefined);
+    }
+  }
+
+  _tap(e: BkEv): void {
+    const renderer = this.model.vertex_renderer;
+    const point = this._map_drag(e.bokeh.sx, e.bokeh.sy, renderer);
+    if (point == null) {
+      return;
+    } else if (this._drawing && this._selected_renderer) {
+      const [x, y] = point;
+      const ds = renderer.data_source;
+      // Type once dataspecs are typed
+      const glyph: any = renderer.glyph;
+      const [xkey, ykey] = [glyph.x.field, glyph.y.field];
+      const index = ds.selected['1d'].indices[0];
+      ds.selected['1d'].indices = [index+1];
+      if (xkey) {
+        const xs = ds.data[xkey];
+        const nx = xs[index];
+        xs[index] = x;
+        xs.splice(index+1, 0, nx)
+      }
+      if (ykey) {
+        const ys = ds.data[ykey];
+        const ny = ys[index];
+        ys[index] = y;
+        ys.splice(index+1, 0, ny)
+      }
+      ds.change.emit(undefined);
+      this._selected_renderer.data_source.change.emit(undefined);
+      return;
+    }
+    const append = e.srcEvent.shiftKey != null ? e.srcEvent.shiftKey : false;
+    this._select_event(e, append, [renderer]);
+    this._select_event(e, append, this.model.renderers);
+  }
+
+  _remove_vertex(emit: boolean = true): void {
+    if (!this._drawing || !this._selected_renderer) { return; }
+    const renderer = this.model.vertex_renderer;
+    const ds = renderer.data_source;
+    // Type once dataspecs are typed
+    const glyph: any = renderer.glyph;
+    const index = ds.selected['1d'].indices[0];
```

Comment:
just noting that this and presumably others will need updates for @clairetang6 PR

Reply:
Yes, I think there are probably a fair number of changes needed, but I've only scanned the PR. I'd be happy to help resolve those issues after a rebase.

Reply:
I think we merge this first, then fix up with your help in the other PR